From 1b7bfa53a6aacb7cdadd985587ab75fadcdb751a Mon Sep 17 00:00:00 2001
From: Vaibhav Hiremath <hvaibhav@ti.com>
Date: Fri, 13 Feb 2009 15:56:40 +0530
Subject: [PATCH 285/426] backlight driver made worked with DSS2

The backlight driver which is posted to community
was not functional with DSS2, made it work.
---
 arch/arm/mach-omap2/board-omap3evm.c  |   37 ++++++++++++++------------------
 drivers/video/backlight/omap3evm_bl.c |    1 +
 include/linux/i2c/twl4030.h           |   19 +++++++++++++++++
 3 files changed, 36 insertions(+), 21 deletions(-)

diff --git a/arch/arm/mach-omap2/board-omap3evm.c b/arch/arm/mach-omap2/board-omap3evm.c
index 0cf7509..1b6be6e 100644
--- a/arch/arm/mach-omap2/board-omap3evm.c
+++ b/arch/arm/mach-omap2/board-omap3evm.c
@@ -268,11 +268,6 @@ static struct platform_device omap3evm_vout_device = {
 #define ENABLE_VPLL2_DEDICATED	0x05
 #define ENABLE_VPLL2_DEV_GRP	0xE0
 
-#define TWL4030_GPIODATA_IN3	0x03
-#define TWL4030_GPIODATA_DIR3	0x06
-#define TWL4030_VPLL2_DEV_GRP	0x33
-#define TWL4030_VPLL2_DEDICATED	0x36
-
 static int lcd_enabled;
 static int dvi_enabled;
 
@@ -312,13 +307,9 @@ static void __init omap3_evm_display_init(void)
 	gpio_direction_output(LCD_PANEL_RESB, 0);
 	gpio_direction_output(LCD_PANEL_QVGA, 0);
 
-#define TWL_LED_LEDEN           0x00
-#define TWL_PWMA_PWMAON         0x00
-#define TWL_PWMA_PWMAOFF        0x01
-
-	twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x11, TWL_LED_LEDEN);
-	twl4030_i2c_write_u8(TWL4030_MODULE_PWMA, 0x01, TWL_PWMA_PWMAON);
-	twl4030_i2c_write_u8(TWL4030_MODULE_PWMA, 0x02, TWL_PWMA_PWMAOFF);
+	twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x11, TWL4030_LED_EN);
+	twl4030_i2c_write_u8(TWL4030_MODULE_PWMA, 0x01, TWL4030_LED_PWMON);
+	twl4030_i2c_write_u8(TWL4030_MODULE_PWMA, 0x02, TWL4030_LED_PWMOFF);
 
 	gpio_direction_output(LCD_PANEL_RESB, 1);
 	gpio_direction_output(LCD_PANEL_INI, 1);
@@ -347,9 +338,10 @@ static int omap3_evm_panel_enable_lcd(struct omap_display *display)
 	}
 	if (omap_rev() > OMAP3430_REV_ES1_0) {
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER,
-			ENABLE_VPLL2_DEDICATED, TWL4030_VPLL2_DEDICATED);
+			ENABLE_VPLL2_DEDICATED, TWL4030_PLL2_DEDICATED);
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER,
 			ENABLE_VPLL2_DEV_GRP, TWL4030_VPLL2_DEV_GRP);
+		twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x10, TWL4030_LED_EN);
 	}
 	gpio_direction_output(LCD_PANEL_ENABLE_GPIO, 0);
 	lcd_enabled = 1;
@@ -360,9 +352,10 @@ static void omap3_evm_panel_disable_lcd(struct omap_display *display)
 {
 	if (omap_rev() > OMAP3430_REV_ES1_0) {
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER, 0x0,
-				TWL4030_VPLL2_DEDICATED);
+				TWL4030_PLL2_DEDICATED);
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER, 0x0,
 				TWL4030_VPLL2_DEV_GRP);
+		twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x11, TWL4030_LED_EN);
 	}
 	gpio_direction_output(LCD_PANEL_ENABLE_GPIO, 1);
 	lcd_enabled = 0;
@@ -408,17 +401,18 @@ static int omap3_evm_panel_enable_dvi(struct omap_display *display)
 		printk(KERN_ERR "cannot enable DVI, LCD is enabled\n");
 		return -EINVAL;
 	}
+
 	if (omap_rev() > OMAP3430_REV_ES1_0) {
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER,
-			ENABLE_VPLL2_DEDICATED, TWL4030_VPLL2_DEDICATED);
+			ENABLE_VPLL2_DEDICATED, TWL4030_PLL2_DEDICATED);
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER,
 			ENABLE_VPLL2_DEV_GRP, TWL4030_VPLL2_DEV_GRP);
+		twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x10, TWL4030_LED_EN);
 	}
-
 	twl4030_i2c_write_u8(TWL4030_MODULE_GPIO, 0x80,
-			TWL4030_GPIODATA_IN3);
+			REG_GPIODATADIR1);
 	twl4030_i2c_write_u8(TWL4030_MODULE_GPIO, 0x80,
-			TWL4030_GPIODATA_DIR3);
+			REG_GPIODATAOUT1);
 	dvi_enabled = 1;
 
 	return 0;
@@ -428,15 +422,16 @@ static void omap3_evm_panel_disable_dvi(struct omap_display *display)
 {
 	if (omap_rev() > OMAP3430_REV_ES1_0) {
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER, 0x0,
-				TWL4030_VPLL2_DEDICATED);
+				TWL4030_PLL2_DEDICATED);
 		twl4030_i2c_write_u8(TWL4030_MODULE_PM_RECEIVER, 0x0,
 				TWL4030_VPLL2_DEV_GRP);
+		twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x11, TWL4030_LED_EN);
 	}
 
 	twl4030_i2c_write_u8(TWL4030_MODULE_GPIO, 0x00,
-			TWL4030_GPIODATA_IN3);
+			REG_GPIODATADIR1);
 	twl4030_i2c_write_u8(TWL4030_MODULE_GPIO, 0x00,
-			TWL4030_GPIODATA_DIR3);
+			REG_GPIODATAOUT1);
 	dvi_enabled = 0;
 }
 
diff --git a/drivers/video/backlight/omap3evm_bl.c b/drivers/video/backlight/omap3evm_bl.c
index 9d1b080..c1359c4 100644
--- a/drivers/video/backlight/omap3evm_bl.c
+++ b/drivers/video/backlight/omap3evm_bl.c
@@ -107,6 +107,7 @@ static int omapbl_set_intensity(struct backlight_device *bd)
 	c = ((125 * (100 - intensity)) / 100) + 2;
 
 #if defined(CONFIG_TWL4030_CORE)
+	twl4030_i2c_write_u8(TWL4030_MODULE_LED, 0x11, TWL4030_LED_EN);
 	twl4030_i2c_write_u8(TWL4030_MODULE_PWMA, c, TWL_PWMA_PWMAOFF);
 #endif
 
diff --git a/include/linux/i2c/twl4030.h b/include/linux/i2c/twl4030.h
index 17bd400..7b786a3 100644
--- a/include/linux/i2c/twl4030.h
+++ b/include/linux/i2c/twl4030.h
@@ -84,6 +84,23 @@ int twl4030_i2c_read(u8 mod_no, u8 *value, u8 reg, unsigned num_bytes);
 /*----------------------------------------------------------------------*/
 
 /*
+* LED Module register offsets (use TWL4030_MODULE_LED)
+*/
+
+#define TWL4030_LED_EN			0x0
+
+/*----------------------------------------------------------------------*/
+
+/*
+ * PWMA/B register offsets (use TWL4030_MODULE_PWMA/TWL4030_MODULE_PWMB)
+ */
+#define TWL4030_LED_PWMON		0x0
+#define TWL4030_LED_PWMOFF		0x1
+
+/*----------------------------------------------------------------------*/
+
+
+/*
  * NOTE:  at up to 1024 registers, this is a big chip.
  *
  * Avoid putting register declarations in this file, instead of into
@@ -376,6 +393,8 @@ int twl4030_sih_setup(int module);
 #define TWL4030_VAUX2_DEDICATED		0x1E
 #define TWL4030_VAUX3_DEV_GRP		0x1F
 #define TWL4030_VAUX3_DEDICATED		0x22
+#define TWL4030_VPLL2_DEV_GRP		0x33
+#define TWL4030_PLL2_DEDICATED		0x36
 
 #if defined(CONFIG_TWL4030_BCI_BATTERY) || \
 	defined(CONFIG_TWL4030_BCI_BATTERY_MODULE)
-- 
1.6.2.4

